File: src/main/java/core/Repl.java
package core;

import core.environment.DefaultEnvironment;
import core.environment.Environment;
import core.evaluator.Evaluator;
import core.exceptions.ExInfoException;
import core.exceptions.ThrowableWrapper;
import core.reader.IReader;
import core.reader.Reader;
import core.reader.StringReader;
import core.scm.Error;
import core.scm.InputPort;
import core.scm.OutputPort;
import core.scm.Symbol;
import core.scm.Void;
import core.writer.Writer;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

public class Repl {

  private static final AtomicInteger SYM_COUNTER = new AtomicInteger(0);
  private static final int SYM_LIMIT = 25;

  private static final String WELCOME = "Welcome to Scheme in Java!";
  private static final String PROMPT = "> ";

  private static final Evaluator evaluator = new Evaluator();
  private static final DefaultEnvironment defaultEnvironment = new DefaultEnvironment();

  private static InputPort currentInputPort = new InputPort(new BufferedInputStream(System.in));
  private static OutputPort currentOutputPort = new OutputPort(System.out);

  private static final IReader reader = new Reader(currentInputPort.getInputStream());

  public static void main(String[] args) throws IOException {
    /* Eval lib procedures */
    StringReader stringReader = new StringReader();
    for (String proc : defaultEnvironment.getLibraryProcedures()) {
      for (Object s : stringReader.read(proc)) {
        evaluator.macroexpandAndEvaluate(s, defaultEnvironment);
      }
    }
    repl(WELCOME, PROMPT, defaultEnvironment);
  }

  private static Symbol getNextID() {
    int i = SYM_COUNTER.incrementAndGet();
    if (i == SYM_LIMIT) {
      SYM_COUNTER.set(0);
    }
    return Symbol.intern("$" + i);
  }

  private static void repl(String welcomeMessage, String prompt, Environment env) throws IOException {
    currentOutputPort.writeln(welcomeMessage);
    //noinspection InfiniteLoopStatement
    while (true) {
      try {
        currentOutputPort.write(prompt);
        /* Read and parse a list of S-expressions from Stdin */
        List<Object> sexps = reader.read();
        for (Object expr : sexps) {
          /* Macroexpand and then Evaluate each S-expression */
          Object result = evaluator.macroexpandAndEvaluate(expr, env);
          /* Do not print and do not store void results */
          if (result == Void.VOID) {
            continue;
          }
          /* nil, on the other hand, is a valid result - print it, but not store it */
          if (result == null) {
            currentOutputPort.writeln(Writer.write(result));
            continue;
          }
          /* Put result into environment */
          Symbol id = getNextID();
          env.put(id, result);
          /* Print */
          currentOutputPort.writeln(id + " = " + Writer.write(result));
        }
      } catch (ThrowableWrapper e) {
        /* Unwrap */
        error(e.getCause());
      } catch (Throwable e) {
        error(e);
      }
    }
  }

  private static void error(Throwable e) throws IOException {
    String errorMessage;
    if (e instanceof Error) {
      errorMessage = "Error: " + e.getMessage();
    } else if (e instanceof ExInfoException) {
      errorMessage = e.toString();
    } else {
      StringBuilder sb = new StringBuilder(e.getClass().getSimpleName());
      if (e.getMessage() != null) {
        sb.append(": ").append(e.getMessage());
      }
      StackTraceElement frame = filterStackTrace(e.getStackTrace());
      if (frame != null) {
        sb.append(" (").append(frame.getFileName()).append(':').append(frame.getLineNumber()).append(')');
      }
      errorMessage = sb.toString();
    }
    currentOutputPort.writeln(errorMessage);
  }

  private static StackTraceElement filterStackTrace(StackTraceElement[] stackTraceElements) {
    for (StackTraceElement stackTraceElement : stackTraceElements) {
      if (stackTraceElement.isNativeMethod()) continue;
      String name = stackTraceElement.getClassName();
      if (name.startsWith("sun.reflect") || name.startsWith("java.lang.reflect")) continue;
      return stackTraceElement;
    }
    return null;
  }

  public static InputPort getCurrentInputPort() {
    return currentInputPort;
  }

  public static void setCurrentInputPort(InputPort inputPort) {
    currentInputPort = inputPort;
  }

  public static OutputPort getCurrentOutputPort() {
    return currentOutputPort;
  }

  public static void setCurrentOutputPort(OutputPort outputPort) {
    currentOutputPort = outputPort;
  }
}

File: src/main/java/core/Repl.kt
package core

import core.environment.DefaultEnvironment
import core.environment.Environment
import core.exceptions.ExInfoException
import core.reader.FileReader
import core.reader.Reader
import core.reader.StringReader
import core.scm.Error
import core.scm.InputPort
import core.scm.OutputPort
import core.scm.Symbol
import java.io.BufferedInputStream
import java.io.File
import java.io.IOException
import java.util.concurrent.atomic.AtomicInteger

object Repl {

    private const val RES_SYM_LIMIT = 10
    private const val WELCOME = "Welcome to Scheme in Kotlin!"
    private const val PROMPT = "> "

    private val resCounter = AtomicInteger(0)
    private val environment = DefaultEnvironment()
    private val evaluator = Evaluator(environment).apply {
        with (StringReader()) {
            environment.libraryProcedures.forEach { this@apply.eval(readOne(it)) }
        }
    }

    var currentInputPort  = InputPort(BufferedInputStream(System.`in`))
    var currentOutputPort = OutputPort(System.out)

    internal val reader = Reader(currentInputPort.inputStream)

    private fun getNextResID() = resCounter.incrementAndGet().let {
        if (it == RES_SYM_LIMIT) resCounter.set(0)
        Symbol.intern("$$it")
    }

    @Throws(IOException::class)
    @JvmStatic fun main(args: Array<String>) = when {
        args.isEmpty() -> repl(WELCOME, PROMPT)
        else           -> evaluateFile(args[0])
    }

    /**
     * Read and evaluate a file and then exit
     */
    @Throws(IOException::class)
    private fun evaluateFile(filename: String) = FileReader().read(File(filename)).forEach {
        evaluator.macroexpandAndEvaluate(it)
    }

    /**
     * Run REPL and evaluate user inputs
     */
    @Throws(IOException::class)
    private fun repl(welcomeMessage: String, prompt: String) {
        currentOutputPort.writeln(welcomeMessage)
        while (true) {
            try {
                currentOutputPort.write(prompt)
                /* Read, macroexpand and then Evaluate each S-expression */
                val result = evaluator.macroexpandAndEvaluate(reader.read())
                /* Do not print and do not store void results */
                if (result === Unit) {
                    continue
                }
                /* nil, on the other hand, is a valid result - print it, but not store it */
                if (result == null) {
                    currentOutputPort.writeln(Writer.write(result))
                    continue
                }
                /* Put result into environment */
                val id = getNextResID()
                evaluator.env[id] = result
                /* Print */
                currentOutputPort.writeln("$id = ${Writer.write(result)}")
            } catch (e: Throwable) {
                /* Store caught exception in $e var */
                evaluator.env[Symbol.intern("\$e")] = e
                error(e)
            }
        }
    }

    @Throws(IOException::class)
    private fun error(e: Throwable) = when (e) {
        is Error -> e.toString()
        is ExInfoException -> e.toString()
        else -> StringBuilder(e.javaClass.simpleName).apply {
            e.message?.let { append(": $it") }
            filterStackTrace(e.stackTrace)?.let {
                append(" (").append(it.fileName ?: "NO_SOURCE_FILE").append(':').append(it.lineNumber).append(')')
            }
        }.toString()
    }.apply(currentOutputPort::writeln)

    private fun filterStackTrace(stackTraceElements: Array<StackTraceElement>) = stackTraceElements.firstOrNull {
        !it.isNativeMethod && !it.className.startsWith("sun.reflect") && !it.className.startsWith("java.lang.reflect")
    }
}
